state/src/caching/order_metadata_index.rs

```diff
@@ -54,8 +65,55 @@ impl OrderMetadataIndex {
     ) {
         let (pair, side) = order.pair_and_side();
         let mut index = self.index.write().await;
-        let entry = index.entry(pair).or_insert_with(OrderSideIndex::new);
+        let entry = index.entry(pair.clone()).or_insert_with(OrderSideIndex::new);
         entry.entry(side).or_insert_with(SortedVec::new).insert((matchable_amount, order_id));
+
+        // Update the reverse index
+        let mut reverse_index = self.reverse_index.write().await;
+        reverse_index.insert(order_id, (pair, side));
+    }
+
+    /// Update the matchable amount for an order
+    pub async fn update_matchable_amount(
+        &self,
+        order_id: OrderIdentifier,
+        matchable_amount: Amount,
+    ) {
+        let (pair, side) = self.get_pair_and_side(&order_id).await.unwrap();
+        let mut index = self.index.write().await;
+        let entry = index.entry(pair).or_insert_with(OrderSideIndex::new);
+        let sorted_vec = entry.entry(side).or_insert_with(SortedVec::new);
+
+        // Remove the old entry
+        let idx = sorted_vec.find_index(|(_, oid)| *oid == order_id).unwrap();
+        sorted_vec.remove(idx);
+
+        // Insert the new entry (this will maintain the sort order)
+        sorted_vec.insert((matchable_amount, order_id));
+    }
+
+    /// Remove an order from the index
+    ///
+    /// Note that we do not clean up sub-index entries when their
+    /// lists become empty.
+    pub async fn remove_order(&self, order_id: &OrderIdentifier) -> Option<()> {
+        // Get the pair and side from the reverse index
+        let (pair, side) = self.get_pair_and_side(order_id).await?;
+
+        // Remove from the main index
+        let mut index = self.index.write().await;
+        let side_index = index.get_mut(&pair)?;
+        let sorted_vec = side_index.get_mut(&side)?;
+
+        if let Some(idx) = sorted_vec.find_index(|(_, oid)| oid == order_id) {
+            sorted_vec.remove(idx);
+        }
+
+        // Remove from the reverse index
+        let mut reverse_index = self.reverse_index.write().await;
+        reverse_index.remove(order_id);
```

Comment:
maybe we should move this to the top of the function, after `get_pair_and_side`, since it's infallible. worried that if somehow the main index doesn't contain an order but the reverse index does, we will return early without removing from the reverse index. this _shouldn't_ ever be the case but might as well keep these indices in sync as much as possible

Reply:
good point